Feature bjs2 26232

Silencemess1ah opened the pull request:
- Added Listener
- Redis config and repositories
- tests
- in work : Comment Listener, CommentCache etc.

Silencemess1ah committed:
- v1
- Kafka listener
- Redis config + redis repositories
- tests
- logic on incrementing PostCacheDto likesCount field
- in work CommentCacheRepo listener and etc.

Silencemess1ah committed:
- unused imports

Silencemess1ah commented on src/main/java/faang/school/postservice/repository/redis/PostCacheRepositoryImpl.java:
```diff
@@ -0,0 +1,25 @@
+    private static final String KEY_PREFIX = "Post:";
+
+    @Override
+    public void incrementLikesCount(Long postId) {
+        String key = KEY_PREFIX + postId;
+        BoundHashOperations<String, String, Object> hashOps = redisTemplate.boundHashOps(key);
+        Object currentLikesObj = hashOps.get("likesCount");
+        assert currentLikesObj != null;
+        long currentLikes = ((Number) currentLikesObj).longValue();
+        hashOps.put("likesCount", currentLikes + 1);
+        log.debug("New like counter is : {}", currentLikes + 1);
+    }
```
вот здесь беда, я что только не делал но hashOps.get возвращает Integer который не каститься в Long пришлось пока сделать так в остальном все работает

Silencemess1ah committed:
Merge branch 'unicorn-stream6-news-feed-NPE-Software' into feature-BJS2-26232

# Conflicts:
#	src/main/java/faang/school/postservice/config/properties/kafka/KafkaProperties.java
#	src/main/resources/application-test.yaml
#	src/main/resources/application.yaml
#	src/test/java/faang/school/postservice/util/BaseContextTest.java

Silencemess1ah committed:
- resolved conflicts and fix on redisson client conflict with main @bean

StepnovaVarvara commented on src/main/java/faang/school/postservice/dto/post/PostCacheDto.java:
```diff
@@ -0,0 +1,26 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@RedisHash(value = "Post")
+public class PostCacheDto implements Serializable {
```
Сережа, я эту дто немножко изменила, не забудь поправить у себя

StepnovaVarvara replied:
+ моя задачка по сохранению поста в кеш редиса замержена

StepnovaVarvara reviewed (commented):
Полностью не смотрела, только кусочек. Мб вечером будут силы, сори)

Silencemess1ah committed:
Merge branch 'unicorn-stream6-news-feed-NPE-Software' into feature-BJS2-26232

# Conflicts:
#	src/main/java/faang/school/postservice/config/kafka/KafkaConsumerConfig.java
#	src/main/java/faang/school/postservice/config/redis/RedisConfiguration.java
#	src/main/java/faang/school/postservice/config/redis/RedisProperties.java
#	src/main/resources/application.yaml

Silencemess1ah committed:
- conflicts resolving

Silencemess1ah committed:
- unused imports

Silencemess1ah committed:
- fix test-yaml

mysosnovskaya commented on src/main/java/faang/school/postservice/config/redis/RedisConfiguration.java:
```diff
@@ -65,16 +73,48 @@ public RedisTemplate<String, Object> redisTemplate() {
+    }
+
+    @Bean
+    @Profile("local")
```
Почему только для local?

Silencemess1ah replied:
Ну тут беда была с Redisson конфигом для теста, пришлось делить так, у нас просто стоит такое в yaml поэтому и поставил что окружение как с него, в общем поменяю на прод чтоб путаницы не было и ямл поправлю

mysosnovskaya commented on src/main/java/faang/school/postservice/consumer/like/post/KafkaPostLikeConsumer.java:
```diff
@@ -0,0 +1,50 @@
+
+    private boolean handlePostLikeEvent(Long postId) {
+        String lockKey = "lock:Post:" + postId;
+        RLock lock = redissonClient.getLock(lockKey);
```
Сомневаюсь, что логика с блокировкой должна быть здесь. А если нам где-то в другом месте потребуется вызвать `postCacheRepository.incrementLikesCount(postId);`, то там тоже эту блокировку реализовывать?)

Silencemess1ah replied:
Понял принял

mysosnovskaya commented on src/test/java/faang/school/postservice/repository/cache/PostCacheRepositoryIntegrationTest.java:
```diff
@@ -0,0 +1,66 @@
+    public void whenSaveAndFindMethodsCalledThenSerializeAndDeserializeCorrectlyAndStoreInCache() {
+        postCacheRepository.save(postCacheDto);
+        Optional<PostCacheDto> foundDto = postCacheRepository.findById(postCacheDto.getPostId());
+        if (foundDto.isPresent()) {
```
А если foundDto оказался пустым? То тест будет успешным, но это не то поведение, которое мы ожидаем. Можно так написать:

```
if (foundDto.isEmpty()) {
    fail(); // fails the test
}
long authorId = foundDto.get().getAuthorId();
assertEquals(authorId, 3);
long postId = foundDto.get().getPostId();
assertEquals(postId, 1);
long likesCount = foundDto.get().getLikesCount();
assertEquals(likesCount, 4);
String content = foundDto.get().getContent();
assertEquals(content, "This is test");
```

Silencemess1ah replied:
Да я сначала даже напоролся на это забыл поправить, дто пустая прилетала и тест проходил а второй падал и я не понимал почему null прилетает, оказалось отсюда ноги растут, да так правильней будет

mysosnovskaya commented on src/test/java/faang/school/postservice/util/RedissonTestConfiguration.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Profile;
+
+@TestConfiguration
```
👍

Silencemess1ah committed:
- fixes after review
- locking and incrementing logic move to repository
- test refactor plus small fixes
